fedimint-client/src/lib.rs

```diff
@@ -508,92 +507,137 @@ fn states_add_instance(
     })
 }
 
-/// Atomically-counted ([`Arc`]) handle to [`Client`]
+/// Shared, inner data of [`ClientHandle`]
 ///
-/// Notably it `deref`-s to the [`Client`] where most
-/// methods live.
-#[derive(Debug)]
-pub struct ClientArc {
-    // Use [`ClientArc::new`] instead
+/// The primary responsibility of this struct (and two levels of Arc),
+/// is shutting down `Client` on the drop of last [`ClientHandle`]. See
+/// [`ClientHandleShared::drop`] for details.
+#[derive(Debug, Clone)]
+struct ClientHandleShared {
     inner: Option<Arc<Client>>,
+}
 
-    __use_constructor_to_create: (),
+/// User handle to [`Client`]
+///
+/// Clonable. On the drop of last [`ClientHandle`] the client
+/// will be shut-down, and resources it used freed.
+///
+/// Notably it [`ops::Deref`]s to the [`Client`] where most
+/// methods live.
+#[derive(Debug, Clone)]
+pub struct ClientHandle {
+    inner: Arc<ClientHandleShared>,
 }
 
-impl ClientArc {
+impl ClientHandle {
     /// Create
     fn new(inner: Arc<Client>) -> Self {
-        inner
-            .client_count
-            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
         Self {
-            inner: Some(inner),
-            // this is the constructor
-            __use_constructor_to_create: (),
+            inner: Arc::new(ClientHandleShared {
+                inner: inner.into(),
+            }),
         }
     }
 
     fn as_inner(&self) -> &Arc<Client> {
-        self.inner.as_ref().expect("Inner always set")
+        self.inner.inner.as_ref().expect("Inner always set")
     }
 
     pub async fn start_executor(&self) {
         self.as_inner().start_executor().await
     }
 
-    /// Block for the `client` to no longer contain any strong references.
-    ///
-    /// Some parts of the code can temporarily clone client to perform some
-    /// actions. No further strong references guarantees that the `client`
-    /// is no longer used inside the system.
-    pub async fn wait_until_fully_dropped(self) {
-        let weak = self.downgrade();
-        drop(self);
-
-        for attempt in 0u64.. {
-            if Weak::strong_count(&weak.inner) == 0 {
-                break;
-            }
-            // we want to retry fast, give feedback, but not spam
-            if attempt % 100 == 0 {
-                info!("Waiting for ArcClient to stop being used");
-            }
-            fedimint_core::task::sleep(Duration::from_millis(10)).await;
+    /// Shutdown the client.
+    /// Returns false if there are other clones of [`ClientHandle`].
+    pub async fn shutdown(self) -> bool {
```

Comment:
I feel like this should return a result...

Reply:
maybe Result<(), ClientHandle>

Reply:
Yeap.

BTW. That `Client` with stopped reactor is an extra flag and state, that I'm not sure if it's useful for anything ever. Maybe we should cut it. It's easier if "handle exist == it's running" always holds. @maan2003